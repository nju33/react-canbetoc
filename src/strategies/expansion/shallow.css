[data-canbetoc-toc-hierarchy-level='1'] > a {
  font-weight: bold;
  font-size: var(--react-canbetoc-toc-headin);
}

[data-canbetoc-toc-hierarchy-level='3'],
[data-canbetoc-toc-hierarchy-level='4'],
[data-canbetoc-toc-hierarchy-level='5'],
[data-canbetoc-toc-hierarchy-level='6'],
[data-canbetoc-toc-hierarchy-level='7'],
[data-canbetoc-toc-hierarchy-level='8'],
[data-canbetoc-toc-hierarchy-level='9'],
[data-canbetoc-toc-hierarchy-level='10'] {
  margin-left: var(--react-canbetoc-toc-margin-left, 0.5rem);
}
