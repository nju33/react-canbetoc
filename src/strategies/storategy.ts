import { TraitTocEntry } from '../entities'

export interface TraitStrategy {
  handle: (tocEntries: TraitTocEntry[]) => PromiseLike<void>
}
