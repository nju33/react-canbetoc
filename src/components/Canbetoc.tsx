import { sequenceT } from 'fp-ts/lib/Apply'
import { pipe } from 'fp-ts/lib/function'
import { fold, map, Option, option } from 'fp-ts/lib/Option'
import * as React from 'react'
import { cloneElement, ReactElement, useContext, useMemo } from 'react'
import { createPortal } from 'react-dom'
import {
  createTocClasssNameValue,
  implicitContext,
  StructTocClassNameContext,
  tocClassNameContext
} from '../contexts'
import { TraitStrategy } from '../strategies/strategy'
import { CanbetocList } from './CanbetocList'

export interface CanbetocProps {
  children: React.ReactElement
  id: string
  intersectionObserverInit?: IntersectionObserverInit
  selectors: string[]
  strategy: TraitStrategy
  tocClassName?: Partial<StructTocClassNameContext>
}

export type CanbetocFC = React.FC<Readonly<CanbetocProps>>

export const Canbetoc: CanbetocFC = <Props extends CanbetocProps>({
  children,
  id,
  // The `{}` is just for the purpose to remove `undefined` type
  intersectionObserverInit = {},
  selectors,
  strategy,
  tocClassName
}: Props) => {
  const {
    buildToc,
    getPortable,
    observe,
    observeAttribute,
    tocEntryService
  } = useContext(implicitContext)
  const baseElementRef = React.useRef<Element>(null)
  const tocBaseElementRef = React.useRef<HTMLUListElement>(null)
  const memoizedIntersectionObserverInit = useMemo(() => {
    return {
      // Default rootMargin
      rootMargin: '0px 0px -90%',
      threshold: [0, 0.25, 0.5, 0.75, 1],
      ...intersectionObserverInit
    }
  }, [intersectionObserverInit])

  const [optionStructuredTocEntries, optionTocEntries] = buildToc.useCase(
    baseElementRef,
    selectors
  )
  const optionPortable = getPortable.useCase(id)

  observe.useCase(optionTocEntries, memoizedIntersectionObserverInit)
  observeAttribute.useCase(id, tocBaseElementRef, strategy, optionTocEntries)

  const optionTocElement = useMemo((): Option<ReactElement> => {
    return pipe(
      sequenceT(option)(optionStructuredTocEntries, optionPortable),
      map(([tocEntries, portable]) => {
        const entries = tocEntries.map((entry) =>
          tocEntryService.convertToDaoFrom(entry)
        )

        return createPortal(
          <tocClassNameContext.Provider
            value={createTocClasssNameValue(tocClassName)}>
            <CanbetocList
              ref={tocBaseElementRef}
              tocId={id}
              entries={entries}
            />
          </tocClassNameContext.Provider>,
          portable
        )
      })
    )
  }, [id, optionStructuredTocEntries, optionPortable])

  const cloned = useMemo(() => {
    return cloneElement(children, { ref: baseElementRef })
  }, [children, baseElementRef])

  const result = useMemo(() => {
    return pipe(
      optionTocElement,
      fold(
        () => <>{cloned}</>,
        (tocElement) => {
          return (
            <>
              {tocElement}
              {cloned}
            </>
          )
        }
      )
    )
  }, [optionTocElement, cloned])

  return result
}

Canbetoc.defaultProps = {
  intersectionObserverInit: {}
}
