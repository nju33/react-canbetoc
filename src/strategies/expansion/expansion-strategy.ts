import { pipe } from 'fp-ts/lib/function'
import { map } from 'fp-ts/lib/Option'
import { TraitTocEntry } from '../../entities'
import { TraitStrategy } from '../storategy'

export class ExpansionStrategy implements TraitStrategy {
  async handle(tocEntries: TraitTocEntry[]): Promise<void> {
    tocEntries.forEach((entry) => {
      pipe(
        entry.getElement(),
        map((element) => {
          element.style.height = ''

          Array.from(
            element.querySelectorAll('[data-canbetoc-toc-id]')
          ).forEach((childElement) => {
            if (childElement instanceof HTMLElement) {
              childElement.style.height = ''
            }
          })
        })
      )
    })
  }
}
