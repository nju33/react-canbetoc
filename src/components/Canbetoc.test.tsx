/**
 * @jest-environment jsdom
 */
import 'intersection-observer'
import '@testing-library/jest-dom'
import { act, getAllByRole, render, RenderResult } from '@testing-library/react'
import * as React from 'react'
import { ExpansionStrategy } from '../strategies/expansion'
import { Canbetoc } from './Canbetoc'

describe('<Canbetoc />', () => {
  const navigation = document.createElement('nav')
  navigation.setAttribute('data-canbetoc-id', 'test')
  const container = document.createElement('div')
  let renderResult: RenderResult

  document.body.appendChild(navigation)
  document.body.appendChild(container)

  beforeEach(async () => {
    jest.useFakeTimers()
    renderResult = render(
      <Canbetoc
        id="test"
        selectors={['h2', 'h3']}
        strategy={new ExpansionStrategy()}>
        <div>
          <h2 id="foo">foo</h2>
          <h3 id="bar">bar</h3>
          <h2 id="baz">baz</h2>
        </div>
      </Canbetoc>
    )

    await new Promise((resolve) => {
      setImmediate(resolve, 1000)
    })
  })

  test('snapshot', () => {
    act(() => {
      jest.advanceTimersByTime(100)
    })
    const { asFragment } = renderResult

    expect(asFragment()).toMatchSnapshot()
  })

  describe('elements', () => {
    test('list', () => {
      act(() => {
        jest.advanceTimersByTime(100)
      })
      const [list] = getAllByRole(navigation, 'list')
      expect(list).toHaveClass('react-canbetoc__toc-list')
      expect(list).toHaveClass('react-canbetoc__toc-list--root')
    })

    test('listitem', () => {
      act(() => {
        jest.advanceTimersByTime(100)
      })
      const [list] = getAllByRole(navigation, 'list')
      const listitems = getAllByRole(list, 'listitem')
      const expectListitem = (
        listitem: HTMLElement,
        data: { hierarchyLevel: string; id: string }
      ): void => {
        expect(listitem).toHaveClass('react-canbetoc__toc-list-item')
        expect(listitem).toHaveAttribute(
          'data-canbetoc-toc-active-branch',
          'false'
        )
        expect(listitem).toHaveAttribute(
          'data-canbetoc-toc-active-item',
          'false'
        )
        expect(listitem).toHaveAttribute('data-canbetoc-toc-anchor-height', '0')
        expect(listitem).toHaveAttribute(
          'data-canbetoc-toc-hierarchy-level',
          data.hierarchyLevel
        )
        expect(listitem).toHaveAttribute('data-canbetoc-toc-id', data.id)
        expect(listitem).toHaveAttribute(
          'data-canbetoc-toc-instance-id',
          expect.any(String)
        )
        expect(listitem).toHaveAttribute(
          'data-canbetoc-toc-latest-intersected',
          'false'
        )
      }

      expect(listitems).toHaveLength(3)
      expectListitem(listitems[0], { id: 'foo', hierarchyLevel: '1' })
      expectListitem(listitems[1], { id: 'bar', hierarchyLevel: '2' })
      expectListitem(listitems[2], { id: 'baz', hierarchyLevel: '1' })
    })
  })
})
