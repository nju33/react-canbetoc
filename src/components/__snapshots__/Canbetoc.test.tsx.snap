// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Canbetoc /> snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2
      id="foo"
    >
      foo
    </h2>
    <h3
      id="bar"
    >
      bar
    </h3>
    <h2
      id="baz"
    >
      baz
    </h2>
  </div>
</DocumentFragment>
`;
